### PR TITLE
Python migration

### PR DESCRIPTION
**Pull request by jdplatt**
This change is [Reviewable](https://reviewable.io/reviews/kubeflow/katib/435)

### CODE REVIEW

**Comment by jdplatt** [comment]
Now there are multiple algorithms I decided to refactor the code in places. The main goals were:

1. Move the bayesian_service.py file inside the main package folder so it’s easy to test.
2. Break all the parsing routines in algorithm manager into functions so they can be used independently (e.g. not all algorithms require us to parse the past observations).
3. Push all algorithm specific code (e.g. parsing or setting default parameters) inside the relevant algorithms classes.
4. Move from a long list of function arguments (e.g. when creating an algorithm) to a few objects that mirror similar concepts in the api (e.g. a parameter search space, the algorithm spec).
5. Run both grid, random, and BO services from a single dockerfile.

I was a little confused about the format of the SuggestionSpec for grid search. I looked at the file in the examples folder but was wondering if there is another example somewhere? If not can someone clarify what the structure of the SuggestionSpec for grid search is?

**Comment by jdplatt** [comment]
This addresses #407 .

**Comment by YujiOshima** [comment]
@jdplatt Thank you for your hard work!
We need to rename the suggestion image. Currently, every suggestion is `bayesianoptimization`.
And we are discussing new API https://github.com/kubeflow/katib/pull/427
It affects on suggestion services.
You don't need to change the logic of algorithm but need to modify API.
I think it would be better to pend this until new API is merged.
@jdplatt @johnugeorge WDTY?

**Comment by jdplatt** [comment]
Thanks @YujiOshima! I agree with your plan. How about I rename bayesianoptimization to katib-suggestion? I'll rename the python package at the same time. Do we have a timeline for the new API? It shouldn't take me too long to get update the suggestion code because all contact with the APIs is limited to a few functions.

**Comment by johnugeorge** [comment]
i feel that we should wait for v1alpha2 as manifests have changed.  We might need to create separate directories for v1alpha2 in order to support both versions. 

@jdplatt we expect to merge GRPC API by end of this week. Once GRPC API is merged, I will try to get initial version of v1alpha2 version in 1-2 weeks.   Can you please hold this PR till then and then make changes for v1alpha2 directly? It will be much easier
   Btwn, I see that `grid_search` and `random_search` python files were moved into `pkg/suggestion/bayesianoptimization` Should we keep them in separate directories?

File: scripts/deploy.sh [hunk 1]
```diff
@@ -31,7 +31,7 @@ kubectl apply -f manifests/vizier/ui
 kubectl apply -f manifests/vizier/suggestion/random
 kubectl apply -f manifests/vizier/suggestion/grid
 kubectl apply -f manifests/vizier/suggestion/hyperband
-kubectl apply -f manifests/vizier/suggestion/bayesianoptimization
+kubectl apply -f manifests/vizier/suggestion/katib_suggestion
```
Comment (andreyvelich) [inline scripts/deploy.sh]: Why do we rename folder only for BayesianOptimization?

File: test/e2e/suggestion-config-grid.yml [hunk 2]
```diff
@@ -1,9 +1,6 @@
...
-        name: "DefaultGrid"
-        value: "1"
-    - 
-        name: "learning-rate"
-        value: "4"
-
+  - name: "DefaultGrid"
+    value: "1"
+  - name: "learning-rate"
+    value: "4"
```
Comment (andreyvelich) [inline test/e2e/suggestion-config-grid.yml]: Please, add newline.

**Comment by andreyvelich** [comment]
@jdplatt Thank you for your work! 
I left few comments.

File: pkg/suggestion/katib_suggestion/algorithm_register.py [hunk 3]
```diff
@@ -0,0 +1,8 @@
+from .random_search import RandomSearch
+from .grid_search import GridSearch
+from .bayesian_optimization_algorithm import BOAlgorithm
+
+
+ALGORITHM_REGISTER = {"random_search": RandomSearch,
+                      "grid_search": GridSearch,
+                      "bayesian_optimization": BOAlgorithm}
```
[the remark this answers is not in the document]
Comment (jdplatt) [reply pkg/suggestion/katib_suggestion/algorithm_register.py]: I am, but was worried this PR had already got really large. I plan to do it soon after this is merged.

File: scripts/deploy.sh (hunk 1 above)
Comment (jdplatt) [reply scripts/deploy.sh]: Now that multiple algorithms (some of which aren't BO) sit inside a single image we start renaming bayesianoptimization to katib_suggestion.

**Comment by gaocegege** [comment]
test/e2e/v1alpha2/test-client.go:9:2: no Go files in /mnt/test-data-volume/kubeflow-katib-presubmit-e2e-435-ac68cb5-9088-fe5c/go/src/github.com/kubeflow/katib/pkg/api

**Comment by johnugeorge** [comment]
Can you update the description with the details of restructuring done in this PR ?

File: manifests/v1alpha2/studyjobcontroller/crd.yaml [hunk 4]
```diff
@@ -0,0 +1,19 @@
+apiVersion: apiextensions.k8s.io/v1beta1
```
Comment (johnugeorge) [inline manifests/v1alpha2/studyjobcontroller/crd.yaml]: studyjobcontroller is not needed in v1alpha2. Why are these files added?

**Comment by jdplatt** [comment]
Sure! Along with my changes to the python code to generate suggestions I added v1alpha2 folders for several components. For the following folders I copied my version into a new folder called v1alpha2 and the copied the v1alpha1 folder over from master:

1. manifests
2. scripts
3. pkg/suggestion
4. test/e2e
5. test/scripts

For the cmd/suggestion and test/workflow folders I left the current folders in the same place to avoid creating merge conflicts for other people and added a v1alpha2 folder containing my version. I figured we can clean up the cmd/suggestion folder when we merge the suggestion service into a single image, and the test/workflow folder when we do the integrations tests for v1alpha2. 

Because my v1alpha2 is just the current state of the manifests, there will be need to be more changes as we add the v1alpha2 versions of the different components. For example, if the studyjobcontroller is not needed in v1alpha2 you can delete it in the PR that implements the replacement. This may not be the ideal approach, but this PR was already getting huge and I wanted to get it merged and not have to roll changes for even more components into it.

**Comment by johnugeorge** [comment]
Do you really need to copy all  the files to v1alpha2 folder in this PR? Is there any reason?  I feel that it will be better to add v1alpha2 only for relevant components related to the changes in this PR . eg: manifests for suggestion  

For other components, we can add whenever corresponding v1alpha2 code is added.

**Comment by jdplatt** [comment]
The code for v1alpha1 and v1alpha2 is very similar so it seemed natural to take a copy of v1alpha1 as the starting point of v1alpha2 and make the required alterations, rather than build up v1alpha2 from scratch. I guess I could have created the v1alpha2 folders and cherry-picked only the files I changed to move over, but it seemed easier to move everything at once to make sure nothing was missed.

**Comment by richardsliu** [comment]
@jdplatt We've changed the API implementation quite a bit from v1alpha1 to v1alpha2, so taking a copy of all the files would make it harder for others to determine which files were copied from v1alpha1 and which ones are under development. It also unfortunately makes this PR harder to review. Can we include only the relevant changed files? Or alternatively mark the copied files with a note/comment so we know which files should be removed later?

**Comment by jlewi** [comment]
@jdplatt @richardsliu How's this PR coming? Do we have a goal in mind for getting it merged?

**Comment by richardsliu** [comment]
@jdplatt We should try to get this PR merged soon, otherwise we will be dealing with more conflicts later on. Let me know if you need help with this.

File: manifests/v1alpha2/vizier/suggestion/katib_suggestion/deployment.yaml [hunk 5]
```diff
@@ -0,0 +1,28 @@
+apiVersion: extensions/v1beta1
```
Comment (richardsliu) [inline manifests/v1alpha2/vizier/suggestion/katib_suggestion/deployment.yaml]: Name of the directory should be bayestian-optimization?

File: pkg/suggestion/v1alpha2/katib_suggestion/algorithm_register.py [hunk 6]
```diff
@@ -0,0 +1,8 @@
+from .random_search import RandomSearch
+from .grid_search import GridSearch
+from .bayesian_optimization_algorithm import BOAlgorithm
+
+
+ALGORITHM_REGISTER = {"random_search": RandomSearch,
```
Comment (richardsliu) [inline pkg/suggestion/v1alpha2/katib_suggestion/algorithm_register.py]: How would someone distinguish between an algorithm that is still in development and one that is stable? What if users want to enable or disable a specific algorithm?

Also if a user wants to add a new algorithm, how would they achieve this without changing the shared Docker image?

File: test/scripts/v1alpha2/run-tests.sh [hunk 7]
```diff
@@ -66,6 +66,11 @@ echo "REPO_NAME ${REPO_NAME}"
 echo "VERSION ${VERSION}"
 
 sed -i -e "s@image: katib\/katib-controller@image: ${REGISTRY}\/${REPO_NAME}\/katib-controller:${VERSION}@" manifests/v1alpha2/katib-controller/katib-controller.yaml
+sed -i -e "s@image: katib\/suggestion@image: ${REGISTRY}\/${REPO_NAME}\/suggestion:${VERSION}@" manifests/vizier/suggestion/random/deployment.yaml
```
Comment (richardsliu) [inline test/scripts/v1alpha2/run-tests.sh]: These seem to be missing the v1alpha2 path.

File: manifests/v1alpha2/vizier/suggestion/grid/deployment.yaml [hunk 8]
```diff
@@ -0,0 +1,28 @@
+apiVersion: extensions/v1beta1
```
Comment (richardsliu) [inline manifests/v1alpha2/vizier/suggestion/grid/deployment.yaml]: Is "vizier" necessary here? I think we can just use "suggestion" here. Also going forward we should avoid using the name vizier since it can confuse users.

File: cmd/suggestion/v1alpha2/hyperband/Dockerfile [hunk 9]
```diff
@@ -0,0 +1,10 @@
+FROM golang:alpine AS build-env
+# The GOPATH in the image is /go.
```
Comment (hougangliu) [inline cmd/suggestion/v1alpha2/hyperband/Dockerfile]: we had `cmd/suggestion/hyperband/v1alpha1` already. can you move it in `cmd/suggestion/hyperband/v1alpha2`? do dose others

File: cmd/suggestion/v1alpha2/katib-suggestion/main.py [hunk 10]
```diff
@@ -0,0 +1,26 @@
+import time
+from concurrent import futures
+
+import grpc
+
+from pkg.api.python import api_pb2_grpc
+from pkg.suggestion.katib_suggestion.rpc_service import SuggestionService
```
Comment (hougangliu) [inline cmd/suggestion/v1alpha2/katib-suggestion/main.py]: wrong python path

File: cmd/suggestion/v1alpha2/hyperband/main.go [hunk 11]
```diff
@@ -0,0 +1,27 @@
+package main
```
Comment (hougangliu) [inline cmd/suggestion/v1alpha2/hyperband/main.go]: do you plan to migrate hyperband to python one in another PR? if yes, please move it in this PR

File: test/e2e/v1alpha2/invalid-studyjob.yaml [hunk 12]
```diff
@@ -0,0 +1,26 @@
+apiVersion: "kubeflow.org/v1alpha1"
+kind: StudyJob
```
Comment (hougangliu) [inline test/e2e/v1alpha2/invalid-studyjob.yaml]: I think you need sync up most part in test/ for v1alpha2

File: pkg/suggestion/v1alpha2/katib_suggestion/algorithm_register.py (hunk 6 above)
Comment (hougangliu) [reply pkg/suggestion/v1alpha2/katib_suggestion/algorithm_register.py]: can we make it configurable instead of hard-code

File: pkg/suggestion/v1alpha2/katib_suggestion/rpc_service.py [hunk 13]
```diff
@@ -0,0 +1,102 @@
...
+REQUIRES_PAST_OBSERVATIONS = ["bayesian_optimization", "grid_search"]
+
+
+class SuggestionService(api_pb2_grpc.SuggestionServicer):
+
+    def __init__(self, search_algorithm=SEARCH_ALGORITHM, logger=None):
+        self.manager_addr = MANAGER_ADDRESS
+        self.manager_port = MANAGER_PORT
+        self.logger = logger if (logger is not None) else get_logger()
+        self.search_algorithm = search_algorithm
```
Comment (hougangliu) [inline pkg/suggestion/v1alpha2/katib_suggestion/rpc_service.py]: if search_algorithm is not in ALGORITHM_REGISTER, we should exit with right error message in __init__ instead of exception occurring in line 30 when GetSuggestions.
That is, the corresponding deployment should not be `Running` if search_algorithm is not supported

File: pkg/suggestion/v1alpha2/katib_suggestion/random_search.py [hunk 14]
```diff
@@ -0,0 +1,25 @@
...
+"""
+from .parsing_utils import parse_x_next_vector
+
+
+class RandomSearch:
+
+    def __init__(self, parameter_config, suggestion_config, logger=None):
+        self.parameter_config = parameter_config
+        self.suggestion_config = suggestion_config
+        self.logger = logger
```
Comment (hougangliu) [inline pkg/suggestion/v1alpha2/katib_suggestion/random_search.py]: self.logger unused

File: pkg/suggestion/v1alpha2/katib_suggestion/rpc_service.py [hunk 15]
```diff
@@ -0,0 +1,102 @@
...
+        self.requires_previous = search_algorithm in REQUIRES_PAST_OBSERVATIONS
+
+    def GetSuggestions(self, request, context):
+        suggestion_config = self._parse_suggestion_parameters(request.param_id)
+        suggestion_config = {param.name: param.value for param in suggestion_config}
+        study_conf = self._get_study_config(request.study_id)
```
Comment (hougangliu) [inline pkg/suggestion/v1alpha2/katib_suggestion/rpc_service.py]: study_id field is dropped in v1alpha2

**Comment by hougangliu** [comment]
@jdplatt GetSuggestionsRequest of v1alpha2 changes a lot, I think you need sync up suggestion services of v1alpha2 with it.
Thanks!

**Comment by richardsliu** [comment]
Discussed offline with @jdplatt and @johnugeorge - we (myself and Johnu) will patch this PR and make the necessary fixes based on recent changes to the v1alpha2 API. Thanks John for your contributions.
